wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Transactions.hs

```diff
@@ -0,0 +1,162 @@
+        (F.FilterByPredicate prd q) -> TxMeta.FilterByPredicate (castFilterOrd prd) q
+        (F.FilterByRange q w)       -> TxMeta.FilterByRange q w
+        (F.FilterIn ls)             -> TxMeta.FilterIn ls
+
+castFilterOrd :: F.FilterOrdering -> TxMeta.FilterOrdering
+castFilterOrd pr = case pr of
+    F.Equal            -> TxMeta.Equal
+    F.GreaterThan      -> TxMeta.GreaterThan
+    F.GreaterThanEqual -> TxMeta.GreaterThanEqual
+    F.LesserThan       -> TxMeta.LesserThan
+    F.LesserThanEqual  -> TxMeta.LesserThanEqual
+
+metaToTx :: Kernel.DB -> Word -> Word -> TxMeta -> V1.Transaction
+metaToTx db k current TxMeta{..} =
+    V1.Transaction {
+        txId = V1 _txMetaId,
+        txConfirmations = confirmations,
+        txAmount = V1 _txMetaAmount,
+        txInputs = toPayDistr <$> _txMetaInputs,
+        txOutputs = toPayDistr <$> _txMetaOutputs,
+        txType = if _txMetaIsLocal then V1.LocalTransaction else V1.ForeignTransaction,
+        txDirection = if _txMetaIsOutgoing then V1.OutgoingTransaction else V1.IncomingTransaction,
+        txCreationTime = V1 _txMetaCreationAt,
+        txStatus = status
+    }
+
+        where
+            hdAccountId = HD.HdAccountId (HD.HdRootId $ InDb _txMetaWalletId)
+                                        (HD.HdAccountIx _txMetaAccountId)
+
+            toPayDistr :: (Address, Coin) -> V1.PaymentDistribution
+            toPayDistr (addr, c) = V1.PaymentDistribution (V1 addr) (V1 c)
+
+            mSlot = Kernel.accountTxSlot db hdAccountId _txMetaId
+            isPending = Kernel.accountIsTxPending db hdAccountId _txMetaId
+
+            (status, confirmations) = buildDynamicTxMeta mSlot k current isPending
+
+buildDynamicTxMeta :: Maybe SlotId -> Word -> Word -> Bool -> (V1.TransactionStatus, Word)
+buildDynamicTxMeta mSlot k currentSlot isPending = case isPending of
+    True  -> (V1.Applying, 0)
+    False ->
+        case mSlot of
+        Nothing     -> (V1.WontApply, 0)
+        Just (SlotId (EpochIndex w64) (UnsafeLocalSlotIndex w16)) ->
+            case ((fromIntegral currentSlot) - w64*(fromIntegral k) + (fromIntegral w16) >= fromIntegral k) of -- TODO: fix
+            True  -> (V1.InNewestBlocks, fromIntegral w64) -- TODO: fix
+            False -> (V1.Persisted, fromIntegral w16)      -- TODO: fix
+
+
+-- | We don`t fitler in memory, so totalEntries is unknown, unless TxMeta Database counts them for us.
+-- It is possible due to some error, to have length ls < Page.
+respond :: RequestParams -> [a] -> Maybe Int -> (WalletResponse [a])
+respond RequestParams{..} ls mbTotalEntries =
+    let totalEntries = case mbTotalEntries of
+            Nothing -> 0
+            Just t  -> t
+        PaginationParams{..}  = rpPaginationParams
+        perPage@(PerPage pp)  = ppPerPage
+        currentPage           = ppPage
+        totalPages            = max 1 $ ceiling (fromIntegral totalEntries / (fromIntegral pp :: Double))
```

Comment:
Is this right? Should this be `min`? If this indeed is a bug, let's make sure that the tests can catch it.